node_modules/gh-series/lib/internal/patterns/cambridge/rollover.js

```diff
@@ -0,0 +1,121 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+var moment = require('moment-timezone');
+
+var PatternConstants = require('./constants');
+
+/**
+ * Roll a timestamp over to the next academic year
+ *
+ * This function will determine on what day of what week of which term a timestamp falls and
+ * use that information to return a timestamp that happens on the same day of the same week
+ * of the same term.
+ *
+ * @param  {String}     timestamp   A timestamp to roll over
+ * @param  {Number}     from        The academical year to which the timestamp belongs
+ * @param  {Number}     to          The academical year to which the timestamp should be rolled over
+ * @return {String}                 The rolled over timestamp
+ */
+var rollOverTimestamp = module.exports.rollOverTimestamp = function(timestamp, from, to) {
+    // Convert the timestamp to a momentjs instance as it's easier to work with
+    // We should work in the London timezone as we need to take daylight savings
+    // time into account
+    timestamp = moment.tz(timestamp, 'Europe/London');
+    var time = timestamp.clone();
+
+    // Need to perform date calculations at midnight, otherwise differing times can affect
+    // the day delta between 2 moments.
+    timestamp.hour(0).minute(0).second(0).millisecond(0);
+
+    // Try to figure out which term the timestamp belongs to
+    var termIndex = getTerm(timestamp, from);
+    var termStart = moment.tz(PatternConstants.TERM_DATES[from][termIndex], 'Europe/London');
+
+    // Determine in which week this event took place and what day
+    var dayOfWeek = timestamp.isoWeekday(); // 1 = Monday, 7 = Sunday
+
+    // The dates in the constants start on a Tuesday, but a term "week" really starts on a Thursday
+    var actualStartOfTerm = termStart.add(2, 'day');
+    var days = timestamp.diff(actualStartOfTerm, 'day');
+
+    // Determine in what week the event falls:
+    // If a term starts on Tues the 7th of Oct
+    //  - Week 0 = 7th and 8th of Oct
+    //  - Week 1 = Thurs 9th till 15th
+    var weeks = Math.floor(days / 7);
+
+    // Start with the start date of the term the timestamp belongs to
+    var newTimestamp = moment.tz(PatternConstants.TERM_DATES[to][termIndex], 'Europe/London');
+
+    // Make up for weeks starting on a Thursday
+    newTimestamp.add(2, 'day');
+
+    // Add the necessary amount of weeks
+    newTimestamp.add(weeks, 'week');
+
+    // Ensure the event is hold on the same day. We can't simply use isoWeekday as Cambridge
+    // has to start the week on a Thursday
+    var offset = ((dayOfWeek - actualStartOfTerm.isoWeekday()) + 7) % 7;
+    newTimestamp.add(offset, 'day');
+
+    // Simply copy the time information
+    newTimestamp.hours(time.hours());
+    newTimestamp.minutes(time.minutes());
+    newTimestamp.seconds(time.seconds());
```

Comment:
A lot of it has to do with it just being conceptually easier to reason about. For me, I find it easier to grok this rollover stuff by thinking:
There's a timestamp X in a term of 2014 that needs to be rolled over to a term of 2015.
So the procedure becomes:
1. Find out to which term TS belongs
1.1. Get the start date of the corresponding term in the next year. This will be the starting point Y for rolling X over
2. Find out in which week of the 2014 term the TS belongs
2.1 Add that number of weeks to Y
3. Find out on what day of the week X was
3.1 Ensure Y is on the same day of the week
4. Find out at what time X happened
4.1 Ensure Y happens at the same time

By starting with a new timestamp and setting the time last we also avoid certain DST issues.